client/src/components/PopoverOptionSet/PopoverOptionSet.js

```diff
@@ -0,0 +1,248 @@
+          type="text"
+          value={searchValue}
+        />
+        {this.renderSearchValueClearLink()}
+      </InputGroup>
+    );
+  }
+
+  /**
+   * Render either all buttons available, buttons matching the search term, or a message that
+   * there are no matching buttons
+   * @returns {DOMElement}
+   */
+  renderOptionButtons() {
+    const {
+      buttons, onSearch, provideButtonClickHandler, buttonContainerClassName,
+      emptyResultClassName, buttonClassName, ButtonComponent
+    } = this.props;
+    const { searchValue } = this.state;
+
+    let buttonsToRender = buttons;
+
+    if (searchValue.length !== 0) {
+      buttonsToRender = onSearch(searchValue, buttonsToRender);
+    }
+
+    if (buttonsToRender.length === 0) {
+      return (
+        <div className={classNames(buttonContainerClassName)}>
+          <div className={classNames(emptyResultClassName)}>
+            {i18n._t('PopoverOptionSet.NO_RESULTS', 'No results found')}
+          </div>
+        </div>
+      );
+    }
+
+    return (
+      <div className={classNames(buttonContainerClassName)}>
+        {buttonsToRender.map((button) => (
+          <ButtonComponent
+            {...button.buttonProps}
+            className={
+              classNames(
+                button.className,
+                buttonClassName
+              )
+            }
+            key={button.key}
+            onClick={provideButtonClickHandler(button)}
+          >
+            {button.content}
+          </ButtonComponent>
+        ))}
+      </div>
+    );
+  }
+
+  /**
+   * Render the option set popover
+   * @returns {Popover}
+   */
+  render() {
+    const { container, className, isOpen, placement, target } = this.props;
+
+    return (
+      <Popover
+        className={classNames(className)}
+        container={container}
+        hideArrow
+        isOpen={isOpen}
+        placement={placement}
+        target={target}
+        toggle={this.handleToggle}
+        onKeyDown={this.handleKeyDown}
+      >
+        {this.renderSearchBox()}
+        {this.renderOptionButtons()}
+      </Popover>
+    );
+  }
+}
+
+PopoverOptionSet.propTypes = {
+  buttons: PropTypes.arrayOf(PropTypes.shape({
+    key: PropTypes.string.isRequired,
+    content: PropTypes.node.isRequired,
+    className: PropTypes.oneOfType([
+      PropTypes.string,
+      PropTypes.object,
+      PropTypes.arrayOf(PropTypes.string)
+    ]),
+    buttonProps: PropTypes.object,
+  })).isRequired,
+  // Accepts a function that takes a search term as a first parameter and a set
+  // of buttons to match against that returns a filtered set of buttons
+  // Default search handler assumes button content to be plain text and performs
+  // a simple string.contains check.
+  onSearch: PropTypes.func,
+  // function that accepts a button object & returns an event handler
+  // e.g. (button) => (event) => event.preventDefault() && togglePopover();
+  provideButtonClickHandler: PropTypes.func.isRequired,
```

Comment:
I'm trying to figure out why this needs to be a factory. In doing so, you're guaranteeing that the component is never pure, because that function is dereferenced every time. It's not a huge thing, but I'm wondering:

* Add `onclick` to the `button` shape?
* Have an all-inclusive button handler that passes the `key` to a single function in which the consumer would be expected to step through a `switch` statement.

Reply:
Updated 🙂